### PR TITLE
Optimization iterations

### PR DESCRIPTION
**Pull request by Zolkin1**
I added an argument that lets the user choose the number of optimization iterations to make at each step.

The idea is that you can get better performance by performing the optimization update multiple times at the same state.

The iteration argument gets passed into the controller parameters.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,5 @@ __pycache__
 *.pkl
 *.egg-info
 MUJOCO_LOG.TXT
+
+.idea
```
Comment (vincekurtz) [inline .gitignore:10]: nit: whitespace

Also what is this file type?

File: examples/cart_pole.py [hunk 2]
```diff
@@ -35,6 +35,7 @@
         spline_type="cubic",
         plan_horizon=1.0,
         num_knots=4,
+        iterations=1,
```
Comment (vincekurtz) [inline examples/cart_pole.py:38]: nit: if it's not changing the default behavior, let's not specify parameters that have a reasonable default

File: examples/pusht.py [hunk 3]
```diff
@@ -20,6 +20,7 @@
     plan_horizon=0.5,
     spline_type="zero",
     num_knots=6,
+    iterations=3,
```
Comment (vincekurtz) [inline examples/pusht.py:23]: Should probably note (not sure where is the best place) that while boosting the number of iterations may make the deterministic sim look better, with async or on hardware the resulting delay could make it not worthwhile

File: hydrax/alg_base.py [hunk 4]
```diff
@@ -135,22 +143,33 @@ def optimize(self, state: mjx.Data, params: Any) -> Tuple[Any, Trajectory]:
...
-        knots = jnp.clip(
-            knots, self.task.u_min, self.task.u_max
-        )  # (num_rollouts, num_knots, nu)
-
-        # Roll out the control sequences, applying domain randomizations and
-        # combining costs using self.risk_strategy.
-        rng, dr_rng = jax.random.split(params.rng)
-        rollouts = self.rollout_with_randomizations(
-            state, new_tk, knots, dr_rng
+        def _optimize_loop_body(i: int, carry: Any):
```
Comment (vincekurtz) [inline hydrax/alg_base.py:146]: Have you run tests to confirm that this doesn't add overhead for the default case of iterations = 1?

File: hydrax/alg_base.py [hunk 5]
```diff
@@ -135,22 +143,33 @@ def optimize(self, state: mjx.Data, params: Any) -> Tuple[Any, Trajectory]:
...
+                state, new_tk, knots, dr_rng
+            )
+            params = params.replace(rng=rng)
+
+            # Update the policy parameters based on the combined costs
+            params = self.update_params(params, rollouts)
+
+            return (params, rollouts)
+
+        params, rollouts = _optimize_loop_body(0, (params, None))
```
Comment (vincekurtz) [inline hydrax/alg_base.py:167]: Why do the first iteration separately, rather than a `fori_loop` starting from `0`?

**Review by vincekurtz — CHANGES_REQUESTED** [review]
Nice! This looks like it could be handy, especially for exploring something like offline trajopt with CMA-ES or some of the other evosax tools.

My only real concerns are:
1) Making sure there isn't a performance regression with `iterations = 1`
2) Adding a unit test. Just something simple. You could run `optimize` twice with `iterations = 1`, and then once with `iterations = 2` (and the same seed) and check that they match.

**Comment by Tom0Brien** [comment]
Nice! iirc I believe this will serve as a step towards potential implementation of the dial-mpc style algorithm (https://arxiv.org/abs/2409.15610)

File: .gitignore (hunk 1 above)
Comment (Zolkin1) [reply .gitignore:10]: This is a settings folder that gets generated when working with Jetbrains IDEs.

**Comment by Zolkin1** [comment]
> Nice! iirc I believe this will serve as a step towards potential implementation of a dial-mpc style algorithm (https://arxiv.org/abs/2409.15610)

Yea, that was one of the reasons I thought about adding this.

@vincekurtz it looks like the `num_samples` parameter is in every control algorithm, but not in the base class. I wasn't sure if this was done on purpose.  I now initialize the rollouts parameter to zero manually, but in order to get the correct dimensions I need to now the num samples amount which I currently get through a call to `sample_knots`. Not sure if there is a better way to get this dimension and/or if we can just move `num_samples` to the base class.

Unit test shows that there is a slight difference in timing, but that is to be expected as there is a bit more overhead when calling the function twice.

File: hydrax/alg_base.py [hunk 6]
```diff
@@ -135,22 +143,59 @@ def optimize(self, state: mjx.Data, params: Any) -> Tuple[Any, Trajectory]:
...
-        )  # (num_rollouts, num_knots, nu)
-
-        # Roll out the control sequences, applying domain randomizations and
-        # combining costs using self.risk_strategy.
-        rng, dr_rng = jax.random.split(params.rng)
-        rollouts = self.rollout_with_randomizations(
-            state, new_tk, knots, dr_rng
+
+        # params, rollouts = _optimize_loop_body(0, (params, None))
+
```
Comment (vincekurtz) [inline hydrax/alg_base.py:170]: I think we can avoid all of this pain with pre-allocating `rollouts` of the correct size if we use a `jax.lax.scan` rather than `fori_loop`. In that case, the `carry` in `_optimize_loop_body` should just be `params`. `rollouts` will be the second return argument, which gets stacked. That should be fine, we just need to be sure to return only the rollouts corresponding to the last iteration.

LMK if that's confusing/unclear: I can also try to write it out myself

File: tests/test_opt_iterations.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    params = ctrl.init_params()
+
+    # Jit
+    jit_optimize = jax.jit(ctrl.optimize)
+
+    # Start timer
+    start_time = time.perf_counter()
+
+    jit_optimize(mjx_data, params)
+    jit_optimize(mjx_data, params)
```
Comment (vincekurtz) [inline tests/test_opt_iterations.py:42]: If we want to check that the results match, this should be `params, _ = jit_optimize(mjx_data, params)`, right? Otherwise nice little test!

File: hydrax/alg_base.py (hunk 6 above)
Comment (Zolkin1) [reply hydrax/alg_base.py]: I like that idea - I've made the change.

File: tests/test_opt_iterations.py (hunk 7 above)
Comment (vincekurtz) [reply tests/test_opt_iterations.py:42]: Let's add the check that using 2 iterations gives the same result as running `jit_optimize` twice.

File: hydrax/alg_base.py [hunk 8]
```diff
@@ -135,23 +143,39 @@ def optimize(self, state: mjx.Data, params: Any) -> Tuple[Any, Trajectory]:
...
+            params = self.update_params(params, rollouts)
+
+            return params, rollouts
+
+        params, rollouts = jax.lax.scan(
+            f=_optimize_scan_body, init=params, xs=jnp.arange(self.iterations)
+        )
+
+        # Get the rollouts from the last iteration only
+        rollouts_final = Trajectory(
```
Comment (vincekurtz) [inline hydrax/alg_base.py:171]: Let's use `jax.tree.map` rather than constructing a new trajectory. That way we won't need to update this if the definition of `Trajectory` ever changes in the future

**Review by vincekurtz — CHANGES_REQUESTED** [review]
Two little comments, otherwise looking good! The `scan` approach is definitely much better. I wasn't able to measure any extra overhead for the `iterations = 1` case on my machine.

**Comment by Zolkin1** [comment]
Ok I went through and made those changes - thanks for the suggestions with Jax! I'm definitely still learning all the idiomatic ways to do things in Jax. Lmk if there are any other desired changes.

Btw, I considered adding an __eq__ method to the `SamplingParams` class, but it looks like that would need to be implemented for each subclass because they automatically override the superclass in this case, so I just compared the elements manually.

File: hydrax/alg_base.py [hunk 9]
```diff
@@ -135,23 +143,39 @@ def optimize(self, state: mjx.Data, params: Any) -> Tuple[Any, Trajectory]:
...
-        # Update the policy parameters based on the combined costs
-        params = self.update_params(params, rollouts)
-        return params, rollouts
+        def _get_last_element(leaf: Any):
+            if isinstance(leaf, jax.Array):
+                return leaf[-1]
+            else:
+                return leaf
+
+        rollouts_final = jax.tree.map(_get_last_element, rollouts)
```
Comment (vincekurtz) [inline hydrax/alg_base.py:176]: The idiomatic way to do this is in one line with a lambda function: `rollouts_final = jax.tree.map(lambda x: x[-1], rollouts)`. Otherwise LGTM!

**Review by vincekurtz — APPROVED** [review]
